sweeper+contractcourt: deadline aware in sweeping anchors

yyforyongyu opened the pull request:
Pushing for concept ack before moving forward.

Current design records the deadline in `AnchorResolution`, we can also calculate the deadlines inside [the `sweepAnchor`](https://github.com/lightningnetwork/lnd/blob/master/contractcourt/channel_arbitrator.go#L1090). The former design has the advantage as we can fine control what deadlines are for remote/local commitment transactions, the latter one however can give much smaller code change and retain the deadline logic inside `contractcourt`.

Another thing to consider is when extracting info from mempool is supported, things is likely to change here?

Next step,
- extend the deadline to HTLC claims
- add unit test and itest
- make the sweeper use the deadline (the new conf target), ie, build a new fee estimator


### Some notes on deciding the deadline

For `to_local` and `to_remote` outputs, there are no deadlines needed as they can only be spent by us/them.
In the revocation path, which doesn't need a deadline either as all other paths are CSV locked.

In our local commitment transaction,
- For incoming HTLCs, if we know the preimage, we should spend it before it times out by the remote peer
- For outgoing HTLCs, we are safe as,
  - if things go well, the remote peer reveals the preimage, we can use it to collect the incoming HTLC.
  - if the remote peer disappears, we can time it out.
  - if the remote reveals the preimage after the timeout, we still have time to collect the incoming HTLC as we have a CLTV delta between the incoming/outgoing HTLCs. This not the remote peer's best interest however, as we can take the money via timeout and collect the money using the preimage.

This means the deadline for us would be the least CLTV value found in the incoming HTLCs.

In our remote commitment transaction,
- For incoming HTLCs, it's money we paid to the remote, and it will time out to us when things don't go well.
- For outging HTLCs, it's money given to us, and it's equivalent to the incoming HTLCs found in our local commitment transaction. We can use the preimage to spend them, or they can be timed out by the remote peer.

In this case, the deadline for us would be the least CLTV value found in the outgoing HTLCs. However, since we can only learn the remote commit tx after it's confirmed, there's no point in tracking the deadline here?

Fix #4215

yyforyongyu commented on contractcourt/channel_arbitrator.go:
```diff
@@ -1122,7 +1118,7 @@ func (c *ChannelArbitrator) sweepAnchors(anchors []*lnwallet.AnchorResolution,
 			sweep.Params{
 				Fee: sweep.FeePreference{
-					ConfTarget: anchorSweepConfTarget,
+					ConfTarget: anchor.CommitDeadline,
```
After more thought, the `ConfTarget` should stay as it is and add a new field `Deadline` to the `sweep.Params`. `ConfTarget` essentially expresses the "expected value", while `Deadline` is our end time.

Roasbeef replied:
Not sure I follow, If `Deadline` is set, how would `ConfTarget` be used in practice?

halseth replied:
Alos note that the deadline is absolute block height here, while the conf_target is relative

yyforyongyu replied:
> Not sure I follow, If `Deadline` is set, how would `ConfTarget` be used in practice?

After a second thought, yeah I think they would mean the same thing.

> Alos note that the deadline is absolute block height here, while the conf_target is relative

Yeah I made a mistake here, will change.

Roasbeef commented:
> Another thing to consider is when extracting info from mempool is supported, things is likely to change here?

Do you mean the issue to start watching the mempool for revealed pre-images or something else entirely?

Roasbeef commented:
> In this case, the deadline for us would be the least CLTV value found in the outgoing HTLCs. However, since we can only learn the remote commit tx after it's confirmed, there's no point in tracking the deadline here?

Yeah I think that makes sense, if by remote commitment you mean the latest commitment for the remote party, then IIRC we only _also_ attempt to sweep those anchors as at broadcast time since we don't know what's in the mempool, we dont' know exactly _which_ transaction will confirm in the end. There's also a related proposal to simply re-use the source UTXOs for this case: https://github.com/lightningnetwork/lnd/pull/4860

Roasbeef commented:
> In our local commitment transaction,
> we still have time to collect the incoming HTLC as we have a CLTV delta between the incoming/outgoing HTLCs. T
> This means the deadline for us would be the least CLTV value found in the incoming HTLCs.

What about the inclusion of the CLTV delta into this heuristic? Lets say we have an _outgoing_ HTLC, and the remote peer is asleep at the wheel for w/e reason and won't come online to pull the pre-image (or attempting to redeem the HTLC on chain would actually be a _net loss_). In this case we still want to ensure that the commitment gets confirmed _before_ the CLTV delta starts to "tick", as we want to be able to utilize the full CLTV delta period to actually fully resolve the HTLC on chain.

Roasbeef commented on lnwallet/channel.go:
```diff
@@ -194,6 +194,9 @@ const (
 )
 
+// noCommitDeadline specifies a commitment transaction has no deadline.
+const noCommitDeadline = ^uint32(0)
```
We can also use `math.MaxUint32` here in place.

Roasbeef commented on lnwallet/channel.go:
```diff
@@ -7049,3 +7075,77 @@ func (lc *LightningChannel) unsignedLocalUpdates(remoteMessageIndex,
+// getCommitmentDeadline finds the deadline for a commitment transaction by
+// extracting the minimum CLTV from its HTLCs. For a local commitment
+// transaction, it uses the lease CLTV value from its incoming HTLCs. For a
+// remote commiment transaction, it uses the least CLTV value from the outgoing
```
commiment -> commitment

Roasbeef commented on lnwallet/channel.go:
```diff
@@ -7049,3 +7075,77 @@ func (lc *LightningChannel) unsignedLocalUpdates(remoteMessageIndex,
+				continue
+			}
+
+			deadlines = append(deadlines, htlc.RefundTimeout)
```
Shouldn't deadlines here be a _relative_ value since that's what's fed in as the _confirmation target_ in the next commit?

yyforyongyu replied:
Yeah my oversight

halseth commented on lnwallet/channel.go:
```diff
@@ -7049,3 +7075,77 @@ func (lc *LightningChannel) unsignedLocalUpdates(remoteMessageIndex,
+	}
+
+	// Sort the deadlines ascendingly and return the first one.
+	sort.Slice(deadlines, func(i, j int) bool {
```
nit: can also just keep track of the lowest one when looping through the HTLCs

halseth commented on lnwallet/channel.go:
```diff
@@ -6196,8 +6205,16 @@ func NewLocalForceCloseSummary(chanState *channeldb.OpenChannel,
+	//  - when the chain watcher detects a spend of the commit tx, hence
+	//    it's confirmed, thus we have no deadline for the commit tx.
+	//  - when the channel arbitrator decides to perform a local force
+	//    close, the state would transit from StateBroadcastCommit to
```
since this is the only case (?) the deadline is really needed to be set, can instead move the dealine calculation there (I think the channel arbitrator is already checking HTLC timeouts when deciding to go to chain).

We could also pass the deadline in as a variadic functional argument, so we would only have to pass it in where it was really needed.

yyforyongyu replied:
Yeah if we only need deadline for local force close, which I believe is the case atm, then I don't think here's the best place to put down the logic. The whole `lnwallet` can be unaware of the deadline, and put the calculation in `ChannelArbitrator`, in [`sweepAnchors`](https://github.com/lightningnetwork/lnd/blob/master/contractcourt/channel_arbitrator.go#L1090). Another reason is that, it seems we are remaking the anchor resolutions in `sweepAnchors` anyway, what's made here will be overwrite when we hit `case StateCommitmentBroadcasted`.

yyforyongyu commented:
> What about the inclusion of the CLTV delta into this heuristic? Lets say we have an _outgoing_ HTLC, and the remote peer is asleep at the wheel for w/e reason and won't come online to pull the pre-image (or attempting to redeem the HTLC on chain would actually be a _net loss_). In this case we still want to ensure that the commitment gets confirmed _before_ the CLTV delta starts to "tick", as we want to be able to utilize the full CLTV delta period to actually fully resolve the HTLC on chain.

Yeah for our outgoing HTLCs we still want to go through the timeout path when necessary. I think the deadline logic can be further extended as, in our local commitment tx,

If we have the preimages, the deadline would be the smaller of,
- the least CLTV from the incoming HTLCs which we can use the preimage to claim, or,
- the least CLTV from the outgoing HTLCs.

If we don't have the preimage,our deadline would be the least CLTV from outgoing HTLCs.

yyforyongyu commented:
I've updated to test out the alternative, which is to manage the deadline inside contract court. I think this approach is better as the only place we need a deadline is when we call `sweepAnchors`. Let me know what you think and I'll continue the rest.

yyforyongyu commented on contractcourt/channel_arbitrator.go:
```diff
@@ -1131,11 +1139,102 @@ func (c *ChannelArbitrator) sweepAnchors(anchors []*lnwallet.AnchorResolution,
+// defined to be the smaller of,
+//   - the least CLTV from outgoing HTLCs, or,
+//   - the least CLTV from incoming HTLCs if the preimage is available.
+func (c *ChannelArbitrator) findCommitmentDeadline(
```
This function can also be incorporated into `c.checkCommitChainActions` as we are checking all the HTLCs there anyway. However I think it's best to make its own function for clarity.

Roasbeef replied:
I like this shift to house the logic in the channel arb itself as well, can allow us to unit test the component more discretely.

halseth commented on contractcourt/channel_arbitrator.go:
```diff
@@ -1131,11 +1139,102 @@ func (c *ChannelArbitrator) sweepAnchors(anchors []*lnwallet.AnchorResolution,
+			return err
+		}
+
+		// For the rest anchors, use the default conf target.
```
I'm not sure if this is what we want. If we decided it is urgent that we force close the channel, the same urgency should apply to all anchors.

Roasbeef replied:
IIRC, the current logic (us attempting to anchor down all three) attempts to account for the case that by the time _we _ force close, one of the remote party's commitments is already in the mempool, so then we attempt to try to anchor down that as well. The issue ofc, is that since we don't watch the mempool, we can't tell if they're attempting to anchor it down themselves, by looking at the transaction spending _their_ anchor output.

halseth replied:
> we can only learn that they've been broadcasted after they are confirmed and to be found in a block

We always assume that _one of the three_ valid commitments are in the mempool, but that we don't know which one. Because of this we anchor down all three with the same urgency.

halseth commented on contractcourt/channel_arbitrator.go:
```diff
@@ -1131,11 +1139,102 @@ func (c *ChannelArbitrator) sweepAnchors(anchors []*lnwallet.AnchorResolution,
+
+	minHeight := uint32(math.MaxUint32)
+
+	// We only use the HTLC set from our local commitment transaction.
```
IIUC the HTLC set is the same on all commitments. (I guess they could differ by a single state update)

Roasbeef replied:
Not necessarily,. we have distinct sets as the HTLCs may differ across the commitments. Consider the case where we force close, but the remote party had a dangling commitment with a newly added HTLC, but they never sent us the signature back.

halseth reviewed (changes requested):
Yeah, I like this approach 👍

yyforyongyu commented on lnwallet/channel.go:
```diff
@@ -6358,16 +6358,22 @@ func (lc *LightningChannel) CompleteCooperativeClose(
-// blindly anchor all of these txes down.
+// NewAnchorResolutions returns a slice of anchor resolutions with a length of
+// 3, in which it has,
+//   - at index 0, the anchor for local commitment tx, or nil.
```
Yeah the new struct is better.

I also notice that we always have at least two anchor resolutions here, one for local and the other for remote. Sometimes we would have a third for pending remote. The only reason I made the current change is that we would be able to identify which one is the anchor for local commit tx, with the assumption that **only the local commit tx needs the deadline**.  But if all anchors share the same deadline, seems like the change here is redundant...

halseth replied:
Yeah, I think so. Since if you want to force close the channel, you don't really care which one of the three anchors gets anchored down, as long as one does.

> I also notice that we always have at least two anchor resolutions here,

You could have a single one in the event one side has balance below the dustlimit and no HTLCs, in which case no anchor will manifest itself.

Roasbeef commented on contractcourt/channel_arbitrator.go:
```diff
@@ -1094,7 +1095,14 @@ func (c *ChannelArbitrator) sweepAnchors(anchors []*lnwallet.AnchorResolution,
 	exclusiveGroup := c.cfg.ShortChanID.ToUint64()
 
-	for _, anchor := range anchors {
+	// sweepWithDeadline
```
Unfinished comment?

Roasbeef commented on contractcourt/channel_arbitrator.go:
```diff
@@ -1131,11 +1139,102 @@ func (c *ChannelArbitrator) sweepAnchors(anchors []*lnwallet.AnchorResolution,
+	}
+
+	for _, anchor := range anchors {
+		// The first item is the anchor of the local commitment tx, we
```
Logic will likely change w/ the move to the struct to make things more explicit.

yyforyongyu replied:
Changed.

Roasbeef commented on contractcourt/channel_arbitrator.go:
```diff
@@ -1131,11 +1139,102 @@ func (c *ChannelArbitrator) sweepAnchors(anchors []*lnwallet.AnchorResolution,
+	// Calculate the deadline. Notice that our minHeight may never get
+	// updated, which means the local active HTLCs set is empty. We will
+	// then fall back to the default value.
+	deadline := minHeight - heightHint
```
FWIW, we could make the deadline a bit more forgiving my also incorporating the CLTV delta for a given outgoing HTLC, as even once the HTLC expires, we still have that delta period before it enters race condition territory. However I like having it a bit stricter to start w/, in the end for non time contested HTLCs, users should see _much_ lower chain fees as a result of force close attempts.

Roasbeef replied:
Yeah I think this is fine for now, we can always introduce _more_ slack later.

Roasbeef commented on contractcourt/channel_arbitrator.go:
```diff
@@ -1131,11 +1139,102 @@ func (c *ChannelArbitrator) sweepAnchors(anchors []*lnwallet.AnchorResolution,
+	// then fall back to the default value.
+	deadline := minHeight - heightHint
+	if minHeight == math.MaxUint32 {
+		deadline = anchorSweepConfTarget
```
Given that we now will attempt to enforce a deadline when we actually need to, perhaps we should consider _increasing_ this default conf target. Right now it's 6 blocks, and due to a sort of observer effect and the lack of extra chain awareness of bitcoind's estimator, that results in a rather high fee rate typically.

yyforyongyu replied:
Actually, a second thought that this value could be even larger. Our current design will always use HTLCs to determine a conf target, and this default value is only used when there are no HTLCs in that particular commitment transaction, thus we have no deadlines at all.

halseth replied:
144 sounds good to me. Could always increase it later.

yyforyongyu commented on lnwallet/chainfee/estimator.go:
```diff
@@ -19,7 +19,7 @@ const (
 	// because it's the highest number of confs bitcoind will return a fee
 	// estimate for.
-	maxBlockTarget uint32 = 1009
+	maxBlockTarget uint32 = 1008
```
I think it's wrong all along:(
Remembered seeing this code a while ago and wanted to change it back then. If the target is greater than 1008, `bitcoind` will throw an error msg, sth like,
```
{
    "result": null,
    "error": {
        "code": -8,
        "message": "Invalid conf_target, must be between 1 - 1008"
    },
    "id": null
}
```

yyforyongyu commented:
Working on an itest, meanwhile ready for another look.

halseth commented on contractcourt/channel_arbitrator.go:
```diff
@@ -1135,11 +1146,141 @@ func (c *ChannelArbitrator) sweepAnchors(anchors *lnwallet.AnchorResolutions,
+	}
+
+	// Sweep anchors based on different HTLC sets. Notice the HTLC sets may
+	// differ across commitments, thus their deadline values could vary.
```
I get why you do this, but can they really vary? I thought they had the same CLTV on both commitments always.

yyforyongyu replied:
Yeah I think it's very unlikely. Even there's a diff between remote and local commit txes, I think the local and remote should be the same?
But still, because we have three sets of HTLCs and three commit txes here, structurally we should have three deadlines. If they are the same, then good. If not, maybe something went wrong.

halseth commented on contractcourt/channel_arbitrator.go:
```diff
@@ -1135,11 +1146,141 @@ func (c *ChannelArbitrator) sweepAnchors(anchors *lnwallet.AnchorResolutions,
+			// to us (toLocalHTLCs) is the incoming HTLCs, and
+			// what's sent from us (toRemoteHTLCs) is the outgoing
+			// HTLCs.
+			toLocalHTLCs := htlcs.incomingHTLCs
```
you might have to check this: IIRC `incoming` here means that it pays to us, regardless of whether this is a local or remote commit.

yyforyongyu replied:
yeah further checked [here](https://github.com/lightningnetwork/lnd/blob/master/lnwallet/channel.go#L2645), got it wrong, will fix.

halseth commented on contractcourt/channel_arbitrator.go:
```diff
@@ -1135,11 +1146,141 @@ func (c *ChannelArbitrator) sweepAnchors(anchors *lnwallet.AnchorResolutions,
+	// Calculate the deadline. Notice that our minHeight may never get
+	// updated, which means the passed HTLCs set is empty. We will then
+	// fall back to the default value.
+	deadline := minHeight - heightHint
```
the height hint here is not necessarily the current block is it?

yyforyongyu replied:
I think it could be one block off if a new block is generated during these lines.

What I see here is,
1. new blocks comes in -> `advanceState` -> `stateStep`, which has the current block, [start from this line](https://github.com/lightningnetwork/lnd/blob/master/contractcourt/channel_arbitrator.go#L2166)
2. the other is from [the loop](https://github.com/lightningnetwork/lnd/blob/master/contractcourt/channel_arbitrator.go#L1174) within `advacenState`, when the state is changed from `StateBroadcastCommit` to `StateCommitmentBroadcasted` from [this line](https://github.com/lightningnetwork/lnd/blob/master/contractcourt/channel_arbitrator.go#L926) after we force close the channel. Depending on how fast these get executed, we might have a new block comes in, and we would be back in case 1.

Roasbeef reviewed (approved):
LGTM 🦈

yyforyongyu commented on lntest/itest/lnd_channel_force_close.go:
```diff
@@ -0,0 +1,221 @@
+
+		// We should see only one sweep transaction because the anchor
+		// sweep is skipped.
+		sweepTxns, err := getNTxsFromMempool(
```
Fixed!

yyforyongyu commented on lntest/itest/lnd_channel_force_close.go:
```diff
@@ -0,0 +1,221 @@
+		t.t, int64(feeRateLarge), feeRate, 0.01,
+		"expected fee rate:%d, got fee rate:%d", feeRateLarge, feeRate,
+	)
+
```
Fixed!

yyforyongyu commented on lntest/itest/lnd_channel_force_close.go:
```diff
@@ -0,0 +1,221 @@
+		require.NoError(t.t, err, "htlc mismatch")
+
+		// Alice force closes the channel.
+		_, _, err = net.CloseChannel(ctxt, alice, chanPoint, true)
```
Yeah I usually use `net.Alice` and `net.Bob` whenever possible. It has the benefit that we don't need to wait for the new nodes to be synced. However, iirc, in this test, the nodes need to set the `--protocol.anchors` in order to test the anchor sweep.

yyforyongyu replied:
That being said, I'm planning to refactor the `itest` so that we reuse `net.Alice` and `net.Bob` as much as possible in a different PR.

wpaulino reviewed (approved):
Looks like this is just missing a release notes addition now?

Roasbeef commented:
Hmm, post rebase, all the itests seem to be failing with this error: 
```
        	Error:      	Received unexpected error:
        	            	wanted 2, found 1 txs in mempool: [4c153bb0d464632d8cc4b07ed1a6bbed3b9258de6b957a80ce0286cd7681f402]
        	Test:       	TestLightningNetworkDaemon/01-of-80/neutrino/test_multi-hop_htlc/committype=anchors/local_force_close_immediate_expiry
``` 

This might be due to the change from 6 blocks to 144 blocks? Though fee estimation shouldn't really matter in this setting given that blocks never actually get full.

yyforyongyu commented:
> Hmm, post rebase, all the itests seem to be failing with this error:
> 
> ```
>         	Error:      	Received unexpected error:
>         	            	wanted 2, found 1 txs in mempool: [4c153bb0d464632d8cc4b07ed1a6bbed3b9258de6b957a80ce0286cd7681f402]
>         	Test:       	TestLightningNetworkDaemon/01-of-80/neutrino/test_multi-hop_htlc/committype=anchors/local_force_close_immediate_expiry
> ```
> 
> This might be due to the change from 6 blocks to 144 blocks? Though fee estimation shouldn't really matter in this setting given that blocks never actually get full.

This is actually an edge case I didn't think of. Sorry but you may have to re-review [this commit here](https://github.com/lightningnetwork/lnd/pull/5148/commits/42f6d1d623f335f6ccd4bd61ef73056fcf2cc21a#diff-a0b8064876b1b1d6085fa7ffdbfd38c81cb06c1ca3f34a08dbaacba203cda3ebR1240). In short,
- that `itest` case has a deadline of 0.
- our fee estimator won't allow a conf target of 0.

The deadline logic is now changed to handle the edge case when the sweeping is behind schedule.

yyforyongyu commented:
@Roasbeef @wpaulino New changes need to be reviewed,
- [change the `minBlockTarget` to be 1 for `webAPIEstimator`](https://github.com/lightningnetwork/lnd/pull/5148/commits/d4d15fe5cc10f3a5fdb810702e6cdf565691a101). Curious why did we choose 2 as our min conf target?
- Updated [deadline logic](https://github.com/lightningnetwork/lnd/pull/5148/commits/d436fb4d6ad4cccc12a7723047bd298f589d8086#diff-a0b8064876b1b1d6085fa7ffdbfd38c81cb06c1ca3f34a08dbaacba203cda3ebR1240) to handle the case when we've reached/passed the deadline.
- Update the `itest` to use a [default conf target of 1](https://github.com/lightningnetwork/lnd/pull/5148/commits/8c543303b7b0078b2a078c681b29659ff45273e7).
- Added a release note.

yyforyongyu commented:
`itest` failed with,
```
2021-06-24 14:40:03.146 [ERR] CRTR: Payment 820882106568555ca9ea7c7602b3e6deb88980ab57ee722853361768651d3c67 failed: incorrect_payment_details
```

Running the single test case locally passed. Weird, investigating.

Roasbeef commented:
Saw a test run where `TestBreachDelayedJusticeConfirmation` failed, ran the flake hunter on it locally for a bit and wasn't able to trigger the failure on this branch, nor on master.

Roasbeef commented:
Most of the itest failures should be fixed by #5430, saw this on the latest run though: 
```
    lnd_channel_force_close.go:132: 
        	Error Trace:	lnd_channel_force_close.go:132
        	            				lnd_channel_force_close.go:195
        	            				test_harness.go:112
        	            				lnd_test.go:13878
        	Error:      	Received unexpected error:
        	            	predicate not satisfied after time out
        	Test:       	TestLightningNetworkDaemon/58-of-80/neutrino/commitment_deadline
        	Messages:   	htlc mismatch
```

Roasbeef commented:
> change the minBlockTarget to be 1 for webAPIEstimator. Curious why did we choose 2 as our min conf target?

IIRC, for bitcoind 1 actually maps to 2, and our implementation of the JSON response is sparsely encoded.

yyforyongyu commented:
> Most of the itest failures should be fixed by #5430, saw this on the latest run though:
> 
> ```
>     lnd_channel_force_close.go:132: 
>         	Error Trace:	lnd_channel_force_close.go:132
>         	            				lnd_channel_force_close.go:195
>         	            				test_harness.go:112
>         	            				lnd_test.go:13878
>         	Error:      	Received unexpected error:
>         	            	predicate not satisfied after time out
>         	Test:       	TestLightningNetworkDaemon/58-of-80/neutrino/commitment_deadline
>         	Messages:   	htlc mismatch
> ```

This is now fixed. The travis build still failed due to suspected flakes.

Roasbeef commented on contractcourt/channel_arbitrator.go:
```diff
@@ -1135,11 +1144,129 @@ func (c *ChannelArbitrator) sweepAnchors(anchors *lnwallet.AnchorResolutions,
+		log.Warnf("ChannelArbitrator(%v): deadline is passed with "+
+			"deadlineMinHeight=%d, heightHint=%d",
+			c.cfg.ChanPoint, deadlineMinHeight, heightHint)
+		deadline = 1
```
Hmm curious to see what effect this actually has in the wild...

We'll need to double check that these outputs don't get filtered out from the sweeper due to providing "negative yield". The failure mode I'm thinking off here is something like:
  * We go to chain for some HTLCs 
  * These HTLCs aren't very large 
   * Node is offline for a while for some reason, causes us to miss the deadline 
   * We publish, then try to sweep w/ 100 sat/vb (just an example) 
   * The sweeper filters this out and the sweeps never go through /sadface 

In theory this can already happen w/ smaller HTLCs, which is why we need to more dynamically control what we think is considered dust.

Roasbeef replied:
Actually this is ok since we'll always set the `Force` flag for any anchor sweeps. We def have more follow up work to do in this area, but it's certainly a start and better than the current state of things (no bumping at all, pray to Satoshi).

Roasbeef commented on contractcourt/channel_arbitrator.go:
```diff
@@ -1135,11 +1144,129 @@ func (c *ChannelArbitrator) sweepAnchors(anchors *lnwallet.AnchorResolutions,
+
+	// When the deadline is passed, we will fall back to the smallest conf
+	// target (1 block).
+	case deadlineMinHeight <= heightHint:
```
👍

Roasbeef commented on lntest/itest/lnd_channel_force_close.go:
```diff
@@ -0,0 +1,223 @@
+
+	// We expect the default max fee rate is used. Allow some deviation
+	// because weight estimates during tx generation are estimates.
+	require.InEpsilonf(
```
Nice test function, at times stuff like fees and be pretty fuzzy so I far prefer this to having some brittle arithmetic inline within the test itself.

Roasbeef reviewed (approved):
LGTM 🏆

Just need to make sure the commits build so it can pass one of the CI tests, and this should be good to land. Excellent work tackling this, has been on my list of things we need to address to increase the security of the core routing network. Other worthy follow ups to this include: the fee bumping function itself, updating CLTV values in real-time based on on-chain conditions, and also updating min HTLC and dust settings based on on-chain conditions.

yyforyongyu commented:
> Commit check failing as one of the commits doesn't compile:

This is now fixed after rebase. I think I will make an issue to keep track of what to do next.

Roasbeef commented:
@yyforyongyu I added a new "task" (they like made it an officially PR now I guess?) to https://github.com/lightningnetwork/lnd/issues/4215 that tracks the next step: starting to gradually bump up the fee rate as we approach the deadline in real time (right now we'll just re-query for the current fee for the specified deadline/conf-target). I think it makes sense to make _new_ issues for the dynamic CLTV measures, and the other aspects I mentioned above.

Roasbeef commented:
Some of the tests fail due to a new error in the itests:
```
<time> [ERR] RPCS: [/lnrpc.Lightning/CloseChannel]: rpc error: code = Canceled desc = context canceled
``` 

Which is benign really AFAICT, and is just part of the auto-logging we have for any errors triggered while handling an RPC request.
